final-project/manRoomStates.tex

```diff
@@ -5,21 +5,50 @@ \chapter{Maintaining Room State \\
+every room must have a timestamp associated with it, designating the time it was loaded. Then, in the case of duplication, the timestamps can be compared and the more recently loaded
+instances can be handled appropriately. Following this, clients who were connected to a bad load of a room should be redirected to the home page where they can enter a new flow of room
+creation or room joining.
 
 \begin{figure}[!htb]
   \centering
   \scalebox{0.57}{\includegraphics{Figures/duplicate-rooms.png}}
   \caption{\label{Figure::duplicate-rooms} Correcting system state following duplicate room instances.}
-\end{figure}
\ No newline at end of file
+\end{figure}
+
+\section{Maintaining Room State Across Service Restarts}
+
+When OTT is deployed, the Monolith\index{monolith} is restarted. Also, heroku restarts OTT about every 24 hours. In order to make this not disruptive to end users, room state is constantly being flushed to redis. When OTT starts up, it gets a list of all the rooms that were loaded and tries to restore their state. The problem is that if a monolith restarts, then it will load all the rooms that are in redis. In a deployment with more than one monolith, this results in rooms existing on more than one monolith.
+
+It's possible to simply allow the system to reach equilibrium using the mechanism described in Section \ref{Section::duplicate-rooms-across-monoliths}. However, this would result in a high memory usage on the Monolith\index{monolith} upon startup. It would also result in a lot of unnecessary network traffic and load on redis.
+
+The solution is to use redis sets\cite{redis-sets} to manage the set of rooms currently loaded. This would guarentee that each room gets loaded only once, even if the resulting distribution of rooms is not optimal.
+
+The system needs a mechanism to detect a cold start. This can be done by having all Monoliths touch a key in redis, "monoliths\_present", every second, refreshing the expire time to 2 seconds in the future. If the key expires, then the system is in a cold start. If the system is in a cold start, monoliths that are starting up must start loading rooms from the load\_queue. Otherwise, they must not load any rooms.
+
+However, this scheme implies that "load\_queue" needs to be maintained, and invalidated if a monolith crashes. Instead, each monolith should maintain it's own set of rooms that it has loaded, with the key in the format "rooms\_loaded:UUID", with "UUID" replaced with a valid UUID. The UUID may be random, it does not need to be persistent across restarts. If the system is in a cold start, all monoliths that are starting up must merge all these sets into a new set called "load\_queue", if it does not exist. The per-monolith room sets must expire at the same time room states expire.
+
+\begin{figure}[!h]
+  \centering
+  \includegraphics[width=0.8\textwidth]{Figures/monolith-startup.png}
```

Comment:
Not sure if it's just my laptop, but the figure is cut off on my end 
![image](https://github.com/dyc3/ssw345-group-assignments/assets/72313458/142b5294-c397-4aa7-b8cf-76f03780c85b)